Enable user profile by default

mposolda opened the pull request:
closes #25151

The PR handles the tasks described in the #25151 with the exception of:
- Enable verify-profile action by default (Created follow-up issue #25985 and added to epic #23907 )
- Documentation (tasks already present for the documentation in the epic #23907 )

Few additional notes:
1) This PR also slightly updates `UsernameTemplateMapper` for identity brokering. In some case the claim value from 3rd party identity broker, which is supposed to be mapped to username, is not a simple String claim, but list of values. In this case, the `value` variable in `UsernameTemplateMapper.setUserNameFromTemplate` is a List with single value `foo`. Currently there is call to `value.toString()`, which effectively means that `value` is converted to the String like `[foo]`. This worked fine before, but now with user-profile enabled by default, there is validation that `username` should not contain characters like `[` and `]`. So it means that mapped username value would be invalid. The change in this PR uses simple String value like `foo` instead of `[foo]` in case that 3rd party claim is a List with single value, so the username would be valid according to user-profile validations. The other possibility is to update the test to remove validations from Username, but TBH I think that this change in the `UsernameTemplateMapper` makes more sense, so the username is valid by default.

2) Once we switch to user-profile, we should likely need to document in the Migration guide all the changes for the default validations. For instance we have some tests, which used single character username like `a` (this worked before, but doesn't work now by default due the `length` validator on `username`). It should be documented how can administrators easily remove those validations.

3) Created follow-up tasks for the various things, which I spot when working on this (it is possible those are just documentation issues, but wanted to doublecheck):
- #25979 
- #25984

mposolda committed:
Enable user profile by default
closes #25151

Signed-off-by: mposolda <mposolda@gmail.com>

jonkoops commented:
@mposolda could you take a look at the merge conflicts and failing Cypress tests?

mposolda commented:
> @mposolda could you take a look at the merge conflicts and failing Cypress tests?

@jonkoops Yes, sure. Already fixed some UI test failures, but looks that something is still remaining...

pedroigor commented:
@mposolda Perhaps you want to change this too https://github.com/keycloak/keycloak/blob/main/js/apps/keycloak-server/scripts/start-server.js#L43.

mposolda commented:
> @mposolda Perhaps you want to change this too https://github.com/keycloak/keycloak/blob/main/js/apps/keycloak-server/scripts/start-server.js#L43.

Updated, Thanks! Also fixed the conflict in `EditUser.tsx` file.

jonkoops commented on js/apps/admin-ui/src/realm-settings/GeneralTab.tsx:
```diff
@@ -251,79 +247,40 @@ export const RealmSettingsGeneralTab = ({
-        {isFeatureEnabled(Feature.DeclarativeUserProfile) && (
-          <FormGroup
-            hasNoPaddingTop
-            label={t("userProfileEnabled")}
```
Since this label is no longer used it can be removed from the properties file containing the translations.

jonkoops commented on js/apps/admin-ui/src/realm-settings/GeneralTab.tsx:
```diff
@@ -251,79 +247,40 @@ export const RealmSettingsGeneralTab = ({
-            label={t("userProfileEnabled")}
-            labelIcon={
-              <HelpItem
-                helpText={t("userProfileEnabledHelp")}
```
This label is no longer used either, so it can be removed from the properties as well.

jonkoops commented on js/apps/admin-ui/cypress/e2e/clients_test.spec.ts:
```diff
@@ -28,7 +28,7 @@ const createClientPage = new CreateClientPage();
 const commonPage = new CommonPage();
 const listingPage = new ListingPage();
-const attributesTab = new AttributesTab();
+const attributesTab = new AttributesTab("client");
```
Looking at the constructor of this class it seems this is not required, as the condition only checks for `"user"`. Let's remove this and make the argument optional in the class itself.

jonkoops commented on js/apps/admin-ui/cypress/e2e/group_test.spec.ts:
```diff
@@ -19,7 +19,7 @@ describe("Group test", () => {
   const groupModal = new GroupModal();
   const searchGroupPage = new SearchGroupPage();
-  const attributesTab = new AttributesTab();
+  const attributesTab = new AttributesTab("group");
```
Same here, the `"group"` argument can be omitted.

jonkoops commented on js/apps/admin-ui/cypress/e2e/users_test.spec.ts:
```diff
@@ -143,6 +145,14 @@ describe("User creation", () => {
 
+  it("Select Unmanaged attributes", () => {
+    sidebarPage.goToRealmSettings();
+    cy.wait(1000);
```
We prefer not to do arbitrary waits as it can slow down the tests unnecessarily. Perhaps it would be possible to replace this with a page load wait?
```suggestion
    sidebarPage.waitForPageLoad();
```

jonkoops commented on js/apps/admin-ui/cypress/support/pages/admin-ui/manage/AttributesTab.ts:
```diff
@@ -1,11 +1,25 @@
+  #valueInput: string;
+  #removeBtn: string;
+
+  constructor(context: string) {
```
As mentioned before, since there is only one parameter that is valid we can type this a bit better.
```suggestion
  constructor(context?: "user") {
```

However, perhaps it would make more sense to make this a boolean instead?

```suggestion
  constructor(isForUser = false) {
```

jonkoops commented on js/apps/admin-ui/src/user/UserAttributes.tsx:
```diff
@@ -36,10 +34,10 @@ export const UserAttributes = ({
           })
         }
-        name={isUserProfileEnabled ? "unmanagedAttributes" : "attributes"}
+        name={"unmanagedAttributes"}
```
This doesn't need to be an expression.
```suggestion
        name="unmanagedAttributes"
```

pedroigor reviewed (commented):
@mposolda Nice! We should merge it once the suggestions from UI team are resolved.
